File: flutter/app/lib/candidates/candidate_bar.dart
import 'package:flutter/material.dart';

import 'package:app/engine/engine_controller.dart';
import 'package:app/src/rust/api.dart';

/// 候选栏：拼音缓冲 + 候选每屏 8 个，点击选择；页数>1 时显示 ‹ n/m › 翻页。
/// 无状态：数据与翻页状态均在 EngineController（单一状态源）。
class CandidateBar extends StatelessWidget {
  const CandidateBar({super.key, required this.controller, required this.onTap});

  final EngineController controller;
  final ValueChanged<int> onTap;

  @override
  Widget build(BuildContext context) {
    return ListenableBuilder(
      listenable: controller,
      builder: (context, _) {
        final pageCount = controller.candidatePageCount;
        final candidates = controller.pageCandidates;
        // english 模式：候选栏退化为模式条，切换中/英有明确区域反馈
        if (controller.mode == ApiMode.english) {
          return Container(
            height: 44,
            color: Colors.grey.shade200,
            child: Row(
              children: [
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 12),
                  child: Text(
                    'EN',
                    style: TextStyle(
                      fontSize: 16,
                      fontWeight: FontWeight.bold,
                      color: Colors.blueGrey.shade700,
                    ),
                  ),
                ),
                Text(
                  '字母直接上屏',
                  style: TextStyle(fontSize: 13, color: Colors.grey.shade600),
                ),
              ],
            ),
          );
        }
        return Container(
          height: 44,
          color: Colors.grey.shade200,
          child: Row(
            children: [
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: 12),
                child: Text(
                  controller.buffer,
                  style: const TextStyle(fontSize: 18, color: Colors.black54),
                ),
              ),
              Expanded(
                child: SingleChildScrollView(
                  scrollDirection: Axis.horizontal,
                  child: Row(
                    children: [
                      for (var i = 0; i < candidates.length; i++)
                        _Candidate(
                          candidates[i].text,
                          onTap: () => onTap(i),
                        ),
                      // 拼音无候选：给出可见反馈，避免"打字无反应"错觉
                      if (candidates.isEmpty && controller.buffer.isNotEmpty)
                        Padding(
                          padding: const EdgeInsets.symmetric(horizontal: 12),
                          child: Text(
                            '无匹配',
                            style: TextStyle(
                                fontSize: 13, color: Colors.grey.shade500),
                          ),
                        ),
                    ],
                  ),
                ),
              ),
              if (pageCount > 1) ...[
                IconButton(
                  icon: const Icon(Icons.chevron_left),
                  visualDensity: VisualDensity.compact,
                  onPressed: controller.prevPage,
                ),
                Text(
                  '${controller.candidatePage + 1}/$pageCount',
                  style: const TextStyle(fontSize: 13),
                ),
                IconButton(
                  icon: const Icon(Icons.chevron_right),
                  visualDensity: VisualDensity.compact,
                  onPressed: controller.nextPage,
                ),
              ],
            ],
          ),
        );
      },
    );
  }
}

class _Candidate extends StatelessWidget {
  const _Candidate(this.text, {required this.onTap});

  final String text;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onTap,
      // 16×2 垂直 padding + 28 行高 ≈ 60dp，溢出 44dp 栏高且点按区超可视区
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 8),
        child: Text(text, style: const TextStyle(fontSize: 20)),
      ),
    );
  }
}

File: android/app/src/main/kotlin/io/opi/input/candidate/CandidateBar.kt
package io.opi.input.candidate

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowLeft
import androidx.compose.material.icons.filled.KeyboardArrowRight
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.opi.input.engine.EngineController
import io.opi.input.engine.EngineMode

/**
 * 候选栏：拼音缓冲 + 每屏 8 候选，点击选择；页数>1 时显示 ‹ n/m › 翻页。
 * 无状态：数据与翻页状态均在 EngineController（单一状态源）。
 */
@Composable
fun CandidateBar(controller: EngineController, onTap: (Int) -> Unit) {
    // english 模式：候选栏退化为模式条，切换中/英有明确区域反馈
    if (controller.mode == EngineMode.ENGLISH) {
        Row(
            modifier = Modifier.fillMaxWidth().height(44.dp).background(Color(0xFFEEEEEE)),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                "EN",
                modifier = Modifier.padding(horizontal = 12.dp),
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
                color = Color(0xFF455A64),
            )
            Text("字母直接上屏", fontSize = 13.sp, color = Color(0xFF757575))
        }
        return
    }
    val pageCount = controller.candidatePageCount
    val candidates = controller.pageCandidates
    Row(
        modifier = Modifier.fillMaxWidth().height(44.dp).background(Color(0xFFEEEEEE)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            controller.buffer,
            modifier = Modifier.padding(horizontal = 12.dp),
            fontSize = 18.sp,
            color = Color(0x8A000000),
        )
        Row(
            modifier = Modifier.weight(1f).horizontalScroll(rememberScrollState()),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            for ((i, c) in candidates.withIndex()) {
                Text(
                    c,
                    modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp).clickable { onTap(i) },
                    fontSize = 20.sp,
                )
            }
            // 拼音无候选：给出可见反馈，避免"打字无反应"错觉
            if (candidates.isEmpty() && controller.buffer.isNotEmpty()) {
                Text(
                    "无匹配",
                    modifier = Modifier.padding(horizontal = 12.dp),
                    fontSize = 13.sp,
                    color = Color(0xFF9E9E9E),
                )
            }
        }
        if (pageCount > 1) {
            IconButton(onClick = controller::prevPage) {
                Icon(Icons.Filled.KeyboardArrowLeft, contentDescription = "上一页")
            }
            Text("${controller.candidatePage + 1}/$pageCount", fontSize = 13.sp)
            IconButton(onClick = controller::nextPage) {
                Icon(Icons.Filled.KeyboardArrowRight, contentDescription = "下一页")
            }
        }
    }
}
